aws/data_source_aws_nat_gateway.go

```diff
@@ -0,0 +1,114 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func dataSourceAwsNatGateway() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAwsNatGatewayRead,
+
+		Schema: map[string]*schema.Schema{
+			"id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"state": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"subnet_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"tags": tagsSchemaComputed(),
+
+			"filter": ec2CustomFiltersSchema(),
+		},
+	}
+}
+
+func dataSourceAwsNatGatewayRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	req := &ec2.DescribeNatGatewaysInput{}
+
+	if id, ok := d.GetOk("id"); ok {
+		req.NatGatewayIds = aws.StringSlice([]string{id.(string)})
+	}
+
+	if vpc_id, ok := d.GetOk("vpc_id"); ok {
+		req.Filter = append(req.Filter, buildEC2AttributeFilterList(
+			map[string]string{
+				"vpc-id": vpc_id.(string),
+			},
+		)...)
+	}
+
+	if state, ok := d.GetOk("state"); ok {
+		req.Filter = append(req.Filter, buildEC2AttributeFilterList(
+			map[string]string{
+				"state": state.(string),
+			},
+		)...)
+	}
+
+	if subnet_id, ok := d.GetOk("subnet_id"); ok {
+		req.Filter = append(req.Filter, buildEC2AttributeFilterList(
+			map[string]string{
+				"subnet-id": subnet_id.(string),
+			},
+		)...)
+	}
+
+	req.Filter = append(req.Filter, buildEC2CustomFilterList(
+		d.Get("filter").(*schema.Set),
+	)...)
+	if len(req.Filter) == 0 {
+		// Don't send an empty filters list; the EC2 API won't accept it.
+		req.Filter = nil
+	}
+	log.Printf("[DEBUG] Reading NAT Gateway: %s", req)
+	resp, err := conn.DescribeNatGateways(req)
+	if err != nil {
+		return err
+	}
+	if resp == nil || len(resp.NatGateways) == 0 {
+		return fmt.Errorf("no matching NAT gateway found: %#v", req)
+	}
+	if len(resp.NatGateways) > 1 {
+		return fmt.Errorf("multiple NAT gateways matched; use additional constraints to reduce matches to a single NAT gateway")
+	}
+
+	ngw := resp.NatGateways[0]
+
+	d.SetId(aws.StringValue(ngw.NatGatewayId))
+	d.Set("state", ngw.State)
+	d.Set("subnet_id", ngw.SubnetId)
+	d.Set("vpc_id", ngw.VpcId)
+
+	for _, address := range ngw.NatGatewayAddresses {
+		if *address.AllocationId != "" {
+			d.Set("allocation_id", address.AllocationId)
```

Comment:
@Ninir I tried using this data source and found these attributes are not being included in the output. Running terraform with DEBUG logging I see this warning:

```
2017/11/07 13:04:19 [WARN] Output interpolation "private_ip" failed: Resource 'data.aws_nat_gateway.default' does not have attribute 'private_ip' for variable 'data.aws_nat_gateway.default.private_ip'
```

I do see the values in the AWS response logged by Terraform:

```
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4: <DescribeNatGatewaysResponse xmlns="http://ec2.amazonaws.com/doc/2016-11-15/">
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:     <requestId>1aa2f2a0-c316-4b83-808b-7402d5cfaa32</requestId>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:     <natGatewaySet>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:         <item>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:             <createTime>2017-11-03T23:34:36.000Z</createTime>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:             <natGatewayAddressSet>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:                 <item>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:                     <allocationId>eipalloc-xxxxxx</allocationId>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:                     <networkInterfaceId>eni-xxxxxx</networkInterfaceId>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:                     <privateIp>172.20.8.29</privateIp>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:                     <publicIp>34.237.xyz.xyz</publicIp>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:                 </item>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:             </natGatewayAddressSet>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:             <natGatewayId>nat-xxxxxxxx</natGatewayId>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:             <state>available</state>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:             <subnetId>subnet-xxxxxxxx</subnetId>
2017-11-07T13:04:17.147-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:             <vpcId>vpc-xxxxxxxx</vpcId>
2017-11-07T13:04:17.148-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:         </item>
2017-11-07T13:04:17.148-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4:     </natGatewaySet>
2017-11-07T13:04:17.148-0800 [DEBUG] plugin.terraform-provider-aws_v1.2.0_x4: </DescribeNatGatewaysResponse>
```
Does the list of output-able values have to be included in the Schema definition? https://github.com/terraform-providers/terraform-provider-aws/pull/1294/files#diff-5264f6e572425a4527111c5172455cf6R16

Reply:
Opened https://github.com/terraform-providers/terraform-provider-aws/pull/2209